### ISSUE TITLE
Determine which tensors are weights

### ISSUE THREAD

**Issue by cjolivier01** [pytorch/xla#1721]
## ❓ Questions and Help
Hi, is there a good way to determine which tensors are the output weights from a general training situation such as test_train_mnist.py example within the ir/view/XLATensor/Lowering layer?  

For example, at the point of creation of XLATensor::Data object, or even later at XLATensor::SyncTensorsGraphInternal()?

I have been able to achieve this for data inputs, weight inputs and loss output by scoping a variable from the python layer (data_parallel.py, xa_model.py, etc.), but the weight-out (updated weights into root tuple) seems created asynchronously via pytorch autograd engine, and by the time the code gets into the lowering, I can't seem to find a good way to determine the difference between an arbitrary tensor output and what is to be an output weight (and bias).

Note that I am working at a commit 1b9e49f2db40a6497b0a423ca2a68520c1f73b14 with cherry-pick of the ScopePush stuff.  Are there other things are important to cherry-pick in (presumably things that are possible to cherry-pick in for tf1.14)?

**Comment by dlibenzi** [comment]
I am not sure what you are trying to achieve, but I am currently working on something which might be related:

https://github.com/pytorch/xla/tree/use_user_alias

The purpose of that is so that we can use XLA buffer aliasing to have the results of weight updates, directly deposited onto the weights themselves.
This required a fix on our backend (which we submitted yesterday) to make it work together with the swap/compact we do on the TPU VM side.
That PR is not trying to guess "weight", but instead will be using the tensor's unique ID to match input with outputs and setup aliasing.

**Comment by cjolivier01** [comment]
basically, going into compile, we need to know which items are inputs, which are "real" user-desired outputs and which are weights -- which tensorsd will be xfered every step, because there's no reason to actually transfer the weights off of the device as actual per-step outputs, so they get special handling. check-pointing the weights are done separately in a completely different way. Does that change above facilitate something like this?

**Comment by dlibenzi** [comment]
I don't know what you changed, but in our architecture weights never go back and forth from device to host.
The only time they come back to host is either checkpointing (which is rare event - once in 1000s steps), or if we have an unimplemented op which trips back to pytorch CPU involving weights.

**Comment by cjolivier01** [comment]
the graph needs the modified weights and biases  going into the root tuple. this was achieved naturally by just running one pass on the network.

**Comment by dlibenzi** [comment]
Are you talking about *your* version or the vanilla `pytorch/xla` version?
Because `pytorch/xla` definitely does not transfer weights every time.

**Comment by cjolivier01** [comment]
I haven't really made any functional changes rather than to fudge with test_train_mnist.py to generate a graph that looks how I want it to look:

here, I have weight and bias updates going into tuple.135, as well as the loss -- I am able to filter out the other values (I only really want wegiht/bias/loss into tuple.135 in this example).   However, programatically, from pytorch_xla layer, it seems generalized to the point (which is of course a fantastic thing in general) it seems that I can't tell which of those is the weight and the bias.  Does that make sense?

![image](https://user-images.githubusercontent.com/11234557/75801756-7e7a9e80-5d30-11ea-91a9-327cfe6435d5.png)

**Comment by cjolivier01** [comment]
I am actually filtering out duplicate tensors transferred to the root tuple which occurs in my branch, the reason for which I am uncertain.  i.e. two nodes may have two edges each, pointing to the root tuple.

**Comment by dlibenzi** [comment]
I don't understand why you want to filter out.
We sync all the live tensors, whose value was changed since the last step.
So if you skip those, you will be essentially unrolling the step loop and making the graph change every time (constantly grows).

**Comment by cjolivier01** [comment]
would it be possible to email you directly to explain?

**Comment by cjolivier01** [comment]
My XLA compile is custom (not CPU or GPU).  PyTorch here is hoped to simply build a graph. The compile itself takes care of intermediate/changing values. The entire training occurs on the device, without returning intermediate values, so, for example, only the loss and weight updates are expected to be connected to the root tuple.  other internal tensor updates are handled by the compile.
Imperative behavior of pytorch in this example won't be preserved -- i.e. you won't be able to see intermediate values in the debugger in this case.  

Simiilar simple network from tensorflow (flatten+dense+softmax):

![image](https://user-images.githubusercontent.com/11234557/75921082-5c0e8100-5e15-11ea-8fa0-2a6bf6c0be22.png)
